File: vm/home/tango-cs/training/python-api/writing02.py
#/!/usr/bin/python
#
# client example: reading an attribute , also with simplified syntax
# C. Scafuri - jan. 2016
#

from PyTango import *
import sys

devname = 'sys/tg_test/1'
attname = 'double_spectrum'

#step 1) create the device proxy
try:
    dev = DeviceProxy(devname)
    print 'proxy for %s created' %(devname)
except DevFailed, ex:
        Except.print_exception(ex)
        sys.exit(1) #semplicistic error handling
#prepare value list
value = []
for i in range(0,255,1):
    value.append(i/10.0)
try:
    dev.write_attribute(attname,value)
    print 'write_attribute "%s" OK' % (attname)
except DevFailed, ex:
        Except.print_exception(ex)
        sys.exit(1) #semplicistic error handling

#use numpy.ndarray - must import numpy 
import numpy
arrvalue=numpy.empty((255))
for i in range(0,255,1):
    arrvalue[i]= i/20.0

try:
    dev.write_attribute(attname,arrvalue)
    dev.float_spectrum=arrvalue  #use alternative syntax - on another attribute
    print 'write_attribute "%s" with numpy.ndarray OK' % (attname)
except DevFailed, ex:
        Except.print_exception(ex)
        sys.exit(1) #semplicistic error handling


sys.exit(0)

File: vm/home/tango-cs/examples/training/client-api/writing02.cpp
/*
 * writng02.cpp
 * 
 *  example program for Tango training - writing an attribute spectrum
 * 
 * C. Scafuri
 * Elettra - Sincrotrone Trieste S.C.p.A.
 * 2016
 */



#include <tango.h>
#include <cstdlib>
using namespace Tango;
int main(int argc, char **argv)
{
    string devname = "sys/tg_test/1";
    string attname = "double_spectrum";
    DeviceProxy *dev = 0;
    // step 1) create the device proxy
    try{
        dev = new Tango::DeviceProxy(devname);
    }
    catch(Tango::DevFailed &e)
    {
        Tango::Except::print_exception(e);
        cerr << "failed to create DeviceProxy" << endl;
        exit(EXIT_FAILURE); // trivial error handling

    }
    // prepare DeviceAttribute object!
    vector<DevDouble> values;
    for (unsigned int i=0; i< 256; i++)
        values.push_back(1.1234);

    //DeviceAttribute da(attname,values); //use direct constructor for deviceAttribute
    DeviceAttribute da;
    da.set_name(attname);
    da << values;
    //another way - use float_spectrum_attribute
    
    DevVarFloatArray *fl;
    fl = new DevVarFloatArray();
    fl->length(256);
    float fin=1.24;
    for (unsigned int i=0; i< 256; i++)
        (*fl)[i]=fin;
    DeviceAttribute daf;
    daf.set_name("float_spectrum"); //const char* argument
    daf << fl; //with oveloaded operator

    try{
        dev->write_attribute(da); //write the attribute
        dev->write_attribute(daf); //write the attribute - ORB takes ownership of data - no need to delete
        cout << "write done"<<endl;
    }
    catch(DevFailed &e)
    {
        Tango::Except::print_exception(e);
        cerr << "failed to write attribute" << endl;
        delete dev; // free resources
        exit(EXIT_FAILURE); // trivial error handling
    }
    delete dev; // free resources
    exit(EXIT_SUCCESS);
}